# The effect of the garbage collector

Let's see whether the garbage collector makes any difference
in the speed with which text-fabric loads the BHSA. 

I work on a MacBook Pro (2021) with M1 Pro processor, 32 GB RAM. 

In [1]:
from tf.app import use

# With GC

This is the default.

In [2]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=True)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  0.67s Dataset without structure sections in otext:no structure functions in the T-API
  2.65s All features loaded/computed - for details use TF.isLoaded()
  1.14s All additional features loaded - for details use TF.isLoaded()


In [3]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=True)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  0.77s Dataset without structure sections in otext:no structure functions in the T-API
  2.75s All features loaded/computed - for details use TF.isLoaded()
  1.15s All additional features loaded - for details use TF.isLoaded()


We see three time indications:

task | first time | second time | avg
--- | --- | --- | ---
init | 0.67 | 0.77 | 0.72
feats | 2.65 | 2.75 | 2.70
additional | 1.14 | 1.15 | 1.15




# Without GC

This is not how Text-Fabric runs by default.

In [4]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=False)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  1.18s Dataset without structure sections in otext:no structure functions in the T-API
  3.13s All features loaded/computed - for details use TF.isLoaded()
  1.38s All additional features loaded - for details use TF.isLoaded()


In [5]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=False)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  1.26s Dataset without structure sections in otext:no structure functions in the T-API
  3.23s All features loaded/computed - for details use TF.isLoaded()
  1.15s All additional features loaded - for details use TF.isLoaded()


We see three time indications:

task | first time | second time | avg
--- | --- | --- | ---
init | 1.18 | 1.26 | 1.22
feats | 3.13 | 3.23 | 3.18
additional | 1.38 | 1.15 | 1.27




# Restart and opposite order

Now we restart the notebook and do the same experiments, but in opposite order.

# Without GC

This is not how Text-Fabric runs by default.

In [2]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=False)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  0.67s Dataset without structure sections in otext:no structure functions in the T-API
  2.61s All features loaded/computed - for details use TF.isLoaded()
  1.13s All additional features loaded - for details use TF.isLoaded()


In [3]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=False)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  0.77s Dataset without structure sections in otext:no structure functions in the T-API
  2.74s All features loaded/computed - for details use TF.isLoaded()
  1.14s All additional features loaded - for details use TF.isLoaded()


We see three time indications:

task | first time | second time | avg
--- | --- | --- | ---
init | 0.67 | 0.77 | 0.72
feats | 2.61 | 2.74 | 2.68
additional | 1.13 | 1.14 | 1.14




# With GC

This is the default.

In [4]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=True)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  1.19s Dataset without structure sections in otext:no structure functions in the T-API
  3.16s All features loaded/computed - for details use TF.isLoaded()
  1.39s All additional features loaded - for details use TF.isLoaded()


In [5]:
A = use("ETCBC/bhsa:clone", checkout="clone", silent="verbose", withGc=True)

**Locating corpus resources ...**

This is Text-Fabric 11.2.3
122 features found and 0 ignored
  1.24s Dataset without structure sections in otext:no structure functions in the T-API
  3.20s All features loaded/computed - for details use TF.isLoaded()
  1.14s All additional features loaded - for details use TF.isLoaded()


We see three time indications:

task | first time | second time | avg
--- | --- | --- | ---
init | 1.19 | 1.24 | 1.22
feats | 3.16 | 3.20 | 3.18
additional | 1.39 | 1.14 | 1.27




Here are the average times of the middle task in one table:

when | with GC | without GC
--- | --- | ---
first | 2.70 | 2.68
last | 3.18 | 3.18


# Conclusion

No real performance difference by switching off the garbage collector.